### ISSUE TITLE
Incorrect loss value of huggingface bert example

### ISSUE THREAD

**Issue by pinxuezhao** [pytorch/PiPPy#799]
Hi~ Thanks for your nice repo!
Steps to reproduce the bug:
1. change the original examples/hf/bert/pippy_bert.py to the following:
```python
# Copyright (c) Meta Platforms, Inc. and affiliates
import argparse
import os
from functools import reduce

import torch
from transformers import BertLMHeadModel, BertConfig

import pippy
import pippy.fx
from pippy import run_pippy
from pippy.microbatch import sum_reducer, TensorChunkSpec
from pippy.events import EventsContext
from pippy.hf import PiPPyHFTracer
from pippy.visualizer import events_to_json


pippy.fx.Tracer.proxy_buffer_attributes = True


def get_number_of_params(model):
    return sum(p.numel() for p in model.parameters() if p.requires_grad)


def print_submod_sizes(model_pipe):
    total_params = 0
    for i, sm in enumerate(model_pipe.split_gm.children()):
        params = get_number_of_params(sm)
        print(f"submod_{i} {params // 10 ** 6}M params")
        total_params += params
    print(f"total {total_params // 10 ** 6}M params")


def run_master(_, args):
    print("Using schedule:", args.schedule)

    bert = BertLMHeadModel(BertConfig(is_decoder=True, num_hidden_layers=8, hidden_size=1024, max_position_embeddings=1024, num_attention_heads=16, intermediate_size=4096, attention_probs_dropout_prob=0, hidden_dropout_prob=0))
    print(bert.config)
    print(f"BERT total number of params = {get_number_of_params(bert) // 10 ** 6}M")
    # print(bert)

    chunks = args.chunks or args.world_size
    batches = 100
    bs = 1 * chunks
    seq_length = 1024

    device = args.device
    bert.to(device)

    bert_input_dict = {
        'input_ids': torch.zeros(bs, seq_length, dtype=torch.long, device=device).random_(bert.config.vocab_size),
        'labels': torch.zeros(bs, seq_length, dtype=torch.long, device=device).random_(bert.config.vocab_size),
        'attention_mask': torch.ones(bs, seq_length, device=device)}
    # bert(**bert_input_dict)

    concrete_args = pippy.create_default_args(bert,
                                              except_keys=bert_input_dict.keys())

    output_chunk_spec = {"loss": sum_reducer,
                         "logits": TensorChunkSpec(0)}

    split_policy = pippy.split_into_equal_size(args.world_size)

    print('Instantiating BERT Pipeline')
    pipe_driver = pippy.compile(
        bert,
        num_ranks=args.world_size,
        num_chunks=chunks,
        schedule=args.schedule,
        split_policy=split_policy,
        tracer=PiPPyHFTracer(),
        checkpoint=bool(args.checkpoint),
        output_chunk_spec=output_chunk_spec,
        concrete_args=concrete_args,
    )
    print_submod_sizes(pipe_driver.pipe)

    this_file_name = os.path.splitext(os.path.basename(__file__))[0]

    print('Running BERT pipeline.')
    pipe_visualized_filename = f"{this_file_name}_visualized.json"
    batches_events_contexts = []
    for i in range(batches):
        output = pipe_driver(**bert_input_dict)
        print(output['loss'].cpu())
        batches_events_contexts.append(pipe_driver.retrieve_events())

    all_events_contexts: EventsContext = reduce(lambda c1, c2: EventsContext().update(c1).update(c2),
                                                batches_events_contexts, EventsContext())
    with open(pipe_visualized_filename, "w") as f:
        f.write(events_to_json(all_events_contexts))
    print(f"Saved {pipe_visualized_filename}")
    print('Finished')


if __name__ == "__main__":
    parser = argparse.ArgumentParser()
    parser.add_argument('--world_size', type=int, default=int(os.getenv("WORLD_SIZE", 4)))
    parser.add_argument('--rank', type=int, default=int(os.getenv("RANK", -1)))
    parser.add_argument('--master_addr', type=str, default=os.getenv('MASTER_ADDR', 'localhost'))
    parser.add_argument('--master_port', type=str, default=os.getenv('MASTER_PORT', '29500'))
    parser.add_argument('-s', '--schedule', type=str, default="FillDrain")
    parser.add_argument('--cuda', type=int, default=int(torch.cuda.is_available()))
    parser.add_argument('--checkpoint', type=int, default=0, choices=[0, 1])
    parser.add_argument("--chunks", type=int, default=None)
    args = parser.parse_args()

    run_pippy(run_master, args)
```

The changes include:

	1. change BertConfig (set dropout probs to zero)
	2. set batches = 100 and seq_length = 1024
	3. print the loss of each iteration

2. launch the training with 
```bash
export MASTER_PORT=29500
export MASTER_ADDR=localhost
export CUDA_VISIBLE_DEVICES=0,1,2,3,4,5,6,7
export CUDA_LAUNCH_BLOCKING=1
GPUS_PER_NODE=8
NNODES=1
NODE_RANK=0
DISTRIBUTED_ARGS="--nproc_per_node $GPUS_PER_NODE --nnodes $NNODES --node_rank $NODE_RANK --master_addr $MASTER_ADDR --master_port $MASTER_PORT"

python3 -u -m torch.distributed.run ${DISTRIBUTED_ARGS} pippy_bert.py --checkpoint=0
```
In theory all the printed losses should be the same, but I got nan and inconsistent ones.
<img width="939" alt="image" src="https://github.com/pytorch/PiPPy/assets/106951017/d2fa25db-69e5-474f-afd2-36b745a58d19">
In theory all the printed losses should be the same, but I got nan and inconsistent ones(like 84.2986).

Any suggestions will be appreciated!
